### PR TITLE
log: Overhaul Windows Logger with Pipes and FHs

### LINKED ISSUE #29673: Tee functionality during build broken on Windows
### Steps to reproduce

Install any package on Windows e.g.
```spack install cpuinfo```

Spack's tee functionality does not pipe the output to stdout, only the output file.

### Error message

_No response_

### Information on your system

* **Spack:** 0.17.1-1640-1c8aed358e
* **Python:** 3.8.4
* **Platform:** windows-windows10-AMD64
* **Concretizer:** original

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by johnwparent**
Redo how we handle IO direction for logging on Windows

Previous we closed sys.stdout, replaced it with a new buffer pointing at the logging file, and then dup2'd to redirect stdout. The logfile was then read from a thread, controlled by the main spack thread, and said reader thread forwarded on the logging context to main Spack's dup'd stdout.

This is deeply insufficient for a number of use cases:

It did not support filtering, unicode, control characters, or force_echo

All echo, including force_echo when supported, was racey, due to the disparate mechanisms by which reader thread behavior was controlled vs the context from which it obtained information. I.e. the reader thread could be killed or echo turned off, before any stdout buffers could be flushed (or whatever out-of-sync ordering) and the reader thread wouldn't see the log output it was supposed to echo prior to turning echo off or terminating.

Nested logging, or cached references to stdout and stderr were invalidated. Previously, on io redirection, we were closing stdout and replacing it with our new buffer. On io restoration, we were restored a buffer pointing at console stdout to sys.stdout, but any cached references to stdout (i.e. nested logging contexts or pytest) were invalidated.

Now we:

* do not close whatever sys.stdout refers to, we mearly store a reference to it
* do not construct a new buffer from python's api, instead we perform redirection from the ground up, starting with fhs to fds and compose a buffer from the existing fh's fd.
* Use pipes for communication with our reader thread, which enables filtering, control characters, etc and allows for non racey echo behavior.

At a glance:

* Redirects IO starting at the Windows file handle (fh) level
* Properly maps fhs to fds, and obtains the appropriate buffer wrapping each
* Enables proper subprocess io redirection inheritance
* Manages shared file handles between stderr and stdout on redirect
* Properly redirects IO at each level of handling, kernel, C, and python
* Ensures cached references to console streams are not invalidated
* Ensures correct order of release for buffers -> fhs

Moves toward a unified logging class by introducing fundamental support
 for the concepts leveraged by the non Windows logger classes like pipes

Also enabled logging testing for Windows

Adds a test for nested logging context managers.

### CODE REVIEW

File: lib/spack/spack/llnl/util/tty/log.py [hunk 1]
```diff
@@ -547,67 +554,84 @@ class StreamWrapper:
...
-            self.libc.fflush(self.c_stream)
-        self.sys_stream.flush()
+        # get current system stream for the standard fd we're redirecting
+        sys_stream = getattr(sys, self.sys_attr)
+        try:
+            if sys_stream:
+                # Flush the system stream before redirection
+                sys_stream.flush()
+        except BaseException:
+            # swallow flush errors
```
Comment (scheibelp) [inline lib/spack/spack/llnl/util/tty/log.py:609]: This would indicate mismanagement wouldn't it? We might want to log this at the debug level.

File: lib/spack/spack/llnl/util/tty/log.py [hunk 2]
```diff
@@ -547,67 +554,84 @@ class StreamWrapper:
...
         """Redirect stdout to the given file descriptor."""
-        # Flush the C-level buffer stream
-        if sys.platform.startswith("win32"):
-            self.libc.fflush(None)
-        else:
-            self.libc.fflush(self.c_stream)
-        # Flush and close sys_stream - also closes the file descriptor (fd)
-        sys_stream = getattr(sys, self.sys_attr)
-        sys_stream.flush()
-        sys_stream.close()
```
Comment (scheibelp) [inline lib/spack/spack/llnl/util/tty/log.py:587]: Does this address the case where a context-manager-like functionality above this has saved a reference to sys.stdout?

**Comment by haampie** [comment]
I haven't looked at this in detail, but just wondering if you've seen `class Tee` in `new_installer.py` as well? The `log_output` bits will likely go away with the old installer.

**Comment by johnwparent** [comment]
> I haven't looked at this in detail, but just wondering if you've seen `class Tee` in `new_installer.py` as well? The `log_output` bits will likely go away with the old installer.

Ahh, I have not. Perhaps this PR should be refactored in favor of supporting `Tee` on Windows then. I'm away at the moment, but I'll look at it once I'm back.

Thanks for pointing that out.

**Comment by haampie** [comment]
Let's chat about it on Slack another time :) have a good holidays!

File: lib/spack/spack/test/llnl/util/tty/log.py [hunk 3]
```diff
@@ -165,4 +160,21 @@ def test_log_subproc_and_echo_output(capfd, tmp_path: pathlib.Path):
 
         # Check captured output (echoed content)
         # Note: 'logged' is not echoed because force_echo() scope ended
-        assert capfd.readouterr()[0] == "echo\n"
+        # "print(echo)" above automatically uses an "\r\n" on Windows
```
Comment (scheibelp) [inline lib/spack/spack/test/llnl/util/tty/log.py]: (minor) add Note: or something to indicate this is not a continuation of prior note

File: lib/spack/spack/llnl/util/tty/log.py [hunk 4]
```diff
@@ -547,67 +554,84 @@ class StreamWrapper:
...
-        new_stream = io.TextIOWrapper(new_buffer)
-        setattr(sys, self.sys_attr, new_stream)
-        self.sys_stream = getattr(sys, self.sys_attr)
+        self.flush()
+        # Get fd for new stream
+        redirect_h = write_conn.fileno()
+        dup_redirect_h = dup_fh(redirect_h)
+        # ensure inheritable handles
+        os.set_handle_inheritable(redirect_h, True)
+        # Todo: investigate if this is neccesary if we do the above before the handle dup
```
Comment (scheibelp) [inline lib/spack/spack/llnl/util/tty/log.py]: Question(s): Can this be determined before a merge? I think redirect_h is what's being used by subprocs, and I would have expected that's the only one that needs to be inheritable.

File: lib/spack/spack/llnl/util/tty/log.py [hunk 5]
```diff
@@ -547,67 +554,84 @@ class StreamWrapper:
...
-        self.sys_stream = getattr(sys, self.sys_attr)
-        self.orig_stream_fd = self.sys_stream.fileno()
-        # Save a copy of the original stdout fd in saved_stream
-        self.saved_stream = os.dup(self.orig_stream_fd)
-
-    def redirect_stream(self, to_fd):
+            raise KeyError(self.sys_attr)
+
+        self.saved_stream = getattr(sys, self.sys_attr)
+        self.saved_std_fd = self.saved_stream.fileno()
```
Comment (scheibelp) [inline lib/spack/spack/llnl/util/tty/log.py]: Do we need to track this explicitly? It seems to always be available via `self.saved_stream.fileno()`

If I'm right but you still think it should be tracked, I can think on that. I worry about situations where duplicate data is maintained with separate references though.

File: lib/spack/spack/test/llnl/util/tty/log.py [hunk 6]
```diff
@@ -165,4 +160,21 @@ def test_log_subproc_and_echo_output(capfd, tmp_path: pathlib.Path):
...
+        # and will replace any \n with \r\n (so end=\n does not work)
+        # \r\n is expected and correct here
+        newline = "\r\n" if sys.platform == "win32" else "\n"
```
Comment (scheibelp) [inline lib/spack/spack/test/llnl/util/tty/log.py:171]: It looks like redirect_stream is setting `newline="\n",`, so does that not have any effect?

File: lib/spack/spack/llnl/util/tty/log.py [hunk 7]
```diff
@@ -630,68 +654,107 @@ def __init__(
...
+        except OSError:
+            pass
+        except BaseException as e:
+            tty.error(f"Exception in log writer thread! {e}", stream=stdout)
+            traceback.print_exc(file=stdout)
+        finally:
+            read.close()
+            log_writer.close()
+            stdout.flush()
+            stdout.close()
```
Comment (scheibelp) [inline lib/spack/spack/llnl/util/tty/log.py:747]: I don't think you should have to flush before you close.

File: lib/spack/spack/llnl/util/tty/log.py [hunk 8]
```diff
@@ -630,68 +654,107 @@ def __init__(
...
+        log_writer = open(logfile, mode=write_mode)
+        try:
+            while True:
+                data = read.recv_bytes(maxlength=4096)
+                # import pdb; pdb.set_trace()
+                if not data:
+                    # the pipe is closed or otherwise inaccesible
+                    return
+                norm_data = data.decode(encoding="utf-8", errors="replace")
+                # stdout.write(data.decode())
```
Comment (scheibelp) [inline lib/spack/spack/llnl/util/tty/log.py]: vestigial?

File: lib/spack/spack/llnl/util/tty/log.py [hunk 9]
```diff
@@ -630,68 +654,107 @@ def __init__(
...
+                clean_line, num_controls = control.subn("", norm_data)
+
+                log_writer.write(_strip(clean_line).encode(encoding="utf-8"))
+                log_writer.flush()
+                if echo or force_echo:
+                    output = clean_line
+                    if filter_fn:
+                        output = filter_fn(output)
+                    enc = stdout.encoding
```
Comment (scheibelp) [inline lib/spack/spack/llnl/util/tty/log.py:725]: (minor) Can this check be done outside the loop (and e.g. define a lambda here?

Might be worth printing a warning if the encoding is not utf-8

Also question: is there a reason you can't consume the bytes from the pipe and pass them directly through without detecting/matching encoding?

**Review by scheibelp — COMMENTED** [review]
Not complete, but I have a few preliminary requests and questions

**Review by scheibelp — COMMENTED** [review]
Not complete, but I have some preliminary requests/questions

File: lib/spack/spack/llnl/util/tty/log.py (hunk 1 above)
Comment (johnwparent) [reply lib/spack/spack/llnl/util/tty/log.py:609]: Yes, this would indicate an incorrectly timed flush, added debug logging.

File: lib/spack/spack/test/llnl/util/tty/log.py (hunk 3 above)
Comment (johnwparent) [reply lib/spack/spack/test/llnl/util/tty/log.py]: Oh yeah good point. Thanks.

File: lib/spack/spack/test/llnl/util/tty/log.py (hunk 6 above)
Comment (johnwparent) [reply lib/spack/spack/test/llnl/util/tty/log.py:171]: It's been a little while but IIRC this is an artifact of pytest and capfd. If you look at the check above where we are able to use `\n` consistently in the file output, that is the effect of setting newline in redirect_stream.

File: lib/spack/spack/llnl/util/tty/log.py (hunk 5 above)
Comment (johnwparent) [reply lib/spack/spack/llnl/util/tty/log.py]: I think it's safer to track it explicitly yes. Basically it allows us to have a consistent idea of what fd we're using to replace whatever the std stream happens to be at redirection time without needing to consider the state of saved_stream.

File: lib/spack/spack/llnl/util/tty/log.py (hunk 2 above)
Comment (johnwparent) [reply lib/spack/spack/llnl/util/tty/log.py:587]: Yes, we're only closing IO components owned by this logging redirect instance.

File: lib/spack/spack/llnl/util/tty/log.py (hunk 4 above)
Comment (johnwparent) [reply lib/spack/spack/llnl/util/tty/log.py]: Investigating now.

File: lib/spack/spack/llnl/util/tty/log.py (hunk 7 above)
Comment (johnwparent) [reply lib/spack/spack/llnl/util/tty/log.py:747]: You are correct, removed.

File: lib/spack/spack/llnl/util/tty/log.py (hunk 9 above)
Comment (johnwparent) [reply lib/spack/spack/llnl/util/tty/log.py:725]: Windows love utf-16 (wide), so depending what stdout is, it may be utf-16, which is normal and expected, particularly if we're in nested logging contexts and stdout is a pipe. 

> Also question: is there a reason you can't consume the bytes from the pipe and pass them directly through without detecting/matching encoding?

The encoding we're reading may not match the encoding of the output, without this output is often garbled/unreadable (or mandarin characters, I hit those a lot without this)

Comment (scheibelp) [reply lib/spack/spack/llnl/util/tty/log.py:725]: For purposes of echoing, I agree there has to be decoding of the reader bytes, and encoding when passing to stdout (although possibly not when it's not a tty)

For writing to the log file, could the bytes just be passed through?

i.e. for the `log_writer`, can we just

```
log_writer.write(data)
```

vs.

```
norm_data = data.decode(encoding="utf-8", errors="replace")
clean_line, num_controls = control.subn("", norm_data)
log_writer.write(_strip(clean_line).encode(encoding="utf-8"))
```

File: lib/spack/spack/llnl/util/tty/log.py (hunk 4 above)
Comment (johnwparent) [reply lib/spack/spack/llnl/util/tty/log.py]: They both need to be inheritable as they are both involved in redirection. Because one is managed by the OS, and one by python/c, we need both handles to do the redirection and avoid a double close on a singular handle.

File: lib/spack/spack/test/llnl/util/tty/log.py (hunk 6 above)
Comment (scheibelp) [reply lib/spack/spack/test/llnl/util/tty/log.py:171]: It seems to me that depending on the context, the newline format *might* not be `\r\n` on Windows, do you think it would make sense to allow for either line ending on Windows? I think the point is that whatever the line ending ends up being is up to whatever configured stdout.

File: lib/spack/spack/llnl/util/tty/log.py (hunk 5 above)
Comment (scheibelp) [reply lib/spack/spack/llnl/util/tty/log.py]: Maybe renaming this might help: calling it `saved_std_fd` implies to me there is some different FD, but really we are `dup`ing onto that FD and the number stays the same the whole time, so maybe just `std_fd` (plus some sort of comment to that effect) would be helpful.

File: lib/spack/spack/llnl/util/tty/log.py [hunk 10]
```diff
@@ -547,67 +554,79 @@ class StreamWrapper:
...
-        new_buffer = open(self.orig_stream_fd, "wb")
-        new_stream = io.TextIOWrapper(new_buffer)
-        setattr(sys, self.sys_attr, new_stream)
-        self.sys_stream = getattr(sys, self.sys_attr)
+        self.flush()
+        # Get fd for new stream
+        redirect_h = write_conn.fileno()
+        dup_redirect_h = dup_fh(redirect_h)
+        os.set_handle_inheritable(redirect_h, True)
+        os.set_handle_inheritable(dup_redirect_h, True)
```
Comment (scheibelp) [inline lib/spack/spack/llnl/util/tty/log.py]: Continuing from https://github.com/spack/spack/pull/51791#discussion_r3061209036 (which I have collapsed because it's out of date)

If I comment out this, then `spack unit-test -k log.py` still passes. If I comment out both `set_handle_inheritable` lines, then one test fails. Should the tests be expanded to cover this. I can say at the moment I'm having trouble understanding the need and that a test might help.

File: lib/spack/spack/llnl/util/tty/log.py [hunk 11]
```diff
@@ -630,68 +649,104 @@ def __init__(
...
+        read,
+        logfile: str,
+        stdout: io.TextIOBase,
+        append: bool,
+        echo: bool,
+        filter_fn: Optional[Callable],
+    ):
+        force_echo = False
+
+        # Open both write and reading on logfile
```
Comment (scheibelp) [inline lib/spack/spack/llnl/util/tty/log.py]: Why is reading mode also enabled? I only ever see it written (andI don't think we should be reading from it).

File: lib/spack/spack/llnl/util/tty/log.py (hunk 5 above)
Comment (johnwparent) [reply lib/spack/spack/llnl/util/tty/log.py]: Makes sense to me.

File: lib/spack/spack/llnl/util/tty/log.py (hunk 11 above)
Comment (johnwparent) [reply lib/spack/spack/llnl/util/tty/log.py]: I think this comment is vestigial, we used to be doing both, but now we only write since we're closer to doing actual tee behavior.

File: lib/spack/spack/llnl/util/tty/log.py (hunk 9 above)
Comment (johnwparent) [reply lib/spack/spack/llnl/util/tty/log.py:725]: We still need to process the data for control characters, so we still need to decode (in utf-8), to get `clean_line` (devoid of control characters, cruft, etc), and clean line needs to be encoded back to binary to be written to the out pipe for the log (and stdout). So all we're doing "extra" is ensuring a consistent utf-8 format for the log.

Looking a bit lower, we actually do encode the output intended for stdout as whatever the terminal happens to be.

That makes sense for stdout, but for the logs, which are spack controlled entities, we should probably consistently make those utf-8 right?

File: lib/spack/spack/test/llnl/util/tty/log.py (hunk 6 above)
Comment (johnwparent) [reply lib/spack/spack/test/llnl/util/tty/log.py:171]: > It seems to me that depending on the context, the newline format might not be \r\n on Windows

Very correct. 

>  think the point is that whatever the line ending ends up being is up to whatever configured stdout.

There are actually a few cases where opening files with unknown line endings with `\r\n` can cause double newlines (`\r\n\n`) so its often safer in general for us to open streams with `\n`. 

In general, `newlines` is impacting the translation at write and read time through the python API. So to a degree, it doesn't matter what we do, the caller can specify the line ending they want when they open whatever stream they're interested in for reading. 

Lets say someone has stdout opened for `\r\n`, and we redirect it with our logger. If they were to capture our redirected (and tee'd) stdout, they'd need another `open` call, which they'd be able to specify their encoding with. 
The only case in which `\n` wouldn't be translated to `\r\n` is if they specified an explicit translation, or explicitly turned translation off, in which case the user has made a deliberate choice. 

We can see that here, where in most of the cases of these log tests, we're able to consistently and nicely use `\n`, except with capfd, where `\r\n` has been enforced by the capfd mechanism and our logger does not overwrite that.

We also definitely want any files we're working with to be `\n`. 

Opening our pipe as `\n` is I think the right choice as we're the only readers and we don't bother dealing with newlines directly, but using one line ending everywhere that every single modern program understands allows spack to have more generic, portable cross platform code, and is one fewer place we have to deal with line endings. I'm all for forcing Windows to use `\n`.

File: lib/spack/spack/llnl/util/tty/log.py (hunk 10 above)
Comment (johnwparent) [reply lib/spack/spack/llnl/util/tty/log.py]: I think this was just my mistake, you're right, one does not need to be inherited. 

For full context, I observed similar behavior, but I was thinking there may be an untested edge case where we'd want both inheritable (in which case your request for more testing is right on the money), but I don't think the edge case I'm thinking of is reasonable/possible.

**Review by scheibelp — APPROVED** [review]
